platform/linux-bridge.c

```diff
@@ -0,0 +1,141 @@
+{
+    struct rtnl_handle *rth = get_rtnl_handle();
+    struct {
+        struct nlmsghdr n;
+        struct ifinfomsg ifi;
+        char buf[256];
+    } req = {};
+
+    req.n.nlmsg_len = NLMSG_LENGTH(sizeof req.ifi);
+    req.n.nlmsg_flags = NLM_F_REQUEST | NLM_F_REPLACE;
+    req.n.nlmsg_type = RTM_SETLINK;
+    req.ifi.ifi_family = AF_BRIDGE;
+    req.ifi.ifi_index = ifindex;
+
+    addattr8(&req.n, sizeof(req.buf), IFLA_PROTINFO, state);
+
+    return rtnl_talk(rth, &req.n, 0, 0, NULL, NULL, NULL);
+}
+
+#ifdef RTM_GETVLAN
+static int vlan_update(unsigned ifindex, __u16 vid,
+#ifdef HAVE_EXPERIMENTAL_KERNEL_VID_FID_SID_STATE_RELATION
+                       __u16 fid, __u16 mstid,
+#endif
+                       __u8 state)
+{
+    struct rtnl_handle *rth = get_rtnl_handle();
+    struct {
+        struct nlmsghdr n;
+        struct br_vlan_msg bvm;
+        char buf[256];
+    } req = {};
+    char entry_buf[256] = {};
+    struct rtattr *rta = (void *)entry_buf;
+    struct bridge_vlan_info vlan_info = {
+        .vid = vid,
+        .flags = BRIDGE_VLAN_INFO_ONLY_OPTS,
+    };
+    struct rtattr *nest;
+
+    req.n.nlmsg_len = NLMSG_LENGTH(sizeof req.bvm);
+    req.n.nlmsg_flags = NLM_F_REQUEST | NLM_F_REPLACE;
+    req.n.nlmsg_type = RTM_NEWVLAN;
+    req.bvm.family = AF_BRIDGE;
+    req.bvm.ifindex = ifindex;
+
+    rta->rta_type = BRIDGE_VLANDB_ENTRY;
+    rta->rta_len = RTA_LENGTH(0);
+
+    nest = rta_nest(rta, sizeof(entry_buf), BRIDGE_VLANDB_ENTRY);
+    if (vid)
+        rta_addattr_l(rta, sizeof(entry_buf), BRIDGE_VLANDB_ENTRY_INFO, &vlan_info, sizeof(vlan_info));
+#ifdef HAVE_EXPERIMENTAL_KERNEL_VID_FID_SID_STATE_RELATION
+    if (fid)
+        rta_addattr16(rta, sizeof(entry_buf), BRIDGE_VLANDB_ENTRY_FID, fid);
+    if (mstid)
+        rta_addattr16(rta, sizeof(entry_buf), BRIDGE_VLANDB_ENTRY_SID, mstid);
+#endif
+    if (state)
+        rta_addattr8(rta, sizeof(entry_buf), BRIDGE_VLANDB_ENTRY_STATE, state);
+    rta_nest_end(rta, nest);
+
+    addraw_l(&req.n, sizeof(req.buf), RTA_DATA(rta), RTA_PAYLOAD(rta));
+
+    return rtnl_talk(rth, &req.n, 0, 0, NULL, NULL, NULL);
+}
+#endif
+
+#ifdef HAVE_EXPERIMENTAL_KERNEL_VID_FID_SID_STATE_RELATION
+#else
+__u16 vid2mstid[VID_COUNT] = {};
+#endif
+
+int bridge_port_vlan_configure(unsigned ifindex, __u16 vid, __u16 fid, __u16 mstid)
+{
+    INFO("[linux-bridge] port %d vid %d fid %d mstid %d", ifindex, vid, fid, mstid);
+#ifdef RTM_GETVLAN
+#ifdef HAVE_EXPERIMENTAL_KERNEL_VID_FID_SID_STATE_RELATION
+    return vlan_update(ifindex, vid, fid, mstid, 0);
+#else
+    vid2mstid[vid] = mstid;
+    return 0;
+#endif
+#else
+    return -ENOTSUP;
+#endif
+}
+
+int bridge_port_tree_set_state(unsigned ifindex, __u16 mstid, __u8 state)
+{
+    INFO("[linux-bridge] port %d mstid %d state %d", ifindex, mstid, state);
+    if (mstid == 0)
+        return port_update(ifindex, state);
+#ifdef RTM_GETVLAN
+#ifdef HAVE_EXPERIMENTAL_KERNEL_VID_FID_SID_STATE_RELATION
```

Comment:
curios: where will these get defined?

if possible i would not add compiletime behavior for mstpd;
i am a bit out of touch with things, but could we do a runtime detection of these features?
maybe have a global variable  `static bool g_rtm_get_vlan_supported`?

there are are a few cases [don't know how many] where
1. some people prefer to backport patches onto kernel 5.4 [let's say] because it's an LTS kernel
2. some people may want to run mstpd on older kernels where these RTNL parameters get passed; although the kernel ignores then
  * if the kernel would ignore these anyway, could we maybe incorporate it in the mstpd behavior somehow to detect at runtime?

Reply:
> curios: where will these get defined?
> 
> if possible i would not add compiletime behavior for mstpd;
> i am a bit out of touch with things, but could we do a runtime detection of these features?
> maybe have a global variable `static bool g_rtm_get_vlan_supported`?
> 
> there are are a few cases [don't know how many] where
> 
>     1. some people prefer to backport patches onto kernel 5.4 [let's say] because it's an LTS kernel
> 
>     2. some people may want to run mstpd on older kernels where these RTNL parameters get passed; although the kernel ignores then
> 
> 
>     * if the kernel would ignore these anyway, could we maybe incorporate it in the mstpd behavior somehow to detect at runtime?

This is what our code is doing here https://github.com/mstpd/mstpd/pull/103/files#diff-ec925e194e98ee3a76396b491ac94d2715496852feb14bcfdddeff4f9df8f381R322

Try to subscribe to the RTNLGRP_BRVLAN netlink notifications, and if it fails, assume per-vlan stp support is not available and fall back to the old behavior. So you only need newer headers to compile, but not run (which is why just having our own headers might be okay).

You obviously need the newer kernel to use it though (or in general any kernel that has the appropate per-vlan-state support).

Reply:
can we get commit https://github.com/mstpd/mstpd/pull/103/commits/55985dc99d900bca1e89f942d5188ee002bbdcd4   into a separate PR for review?

i know this may be tedious;
but at least it may be one way forward;
i like the idea of the mechanism of trying to subscribe to RTNLGRP_BRVLAN netlink notifications [with failure handling];

maybe after that, @KanjiMonster can you rebase PR https://github.com/mstpd/mstpd/pull/103/ ? 
if the commits go down in that PR, then it makes things easier [for me] to track;